language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,366 @@
+    pub struct_handles: Vec<StructHandle>,
+    pub function_handles: Vec<FunctionHandle>,
+    pub field_handles: Vec<FieldHandle>,
+    // instantiations are references to a *generic* handle and and its instantiation, whether
+    // partial or complete
+    pub struct_instantiations: Vec<StructInstantiation>,
+    pub function_instantiations: Vec<FunctionInstantiation>,
+    pub field_instantiations: Vec<FieldInstantiation>,
+    // common blob tables
+    pub signatures: SignaturePool,
+    pub identifiers: IdentifierPool,
+    // constants
+    pub byte_array_pool: U8VectorPool,
+    pub address_pool: AddressPool,
+}
+
+// In `CompiledScript` we replace the FunctionDefinition for main(..) with broken up data
+// to set up the call frame and define/verify main signature and arguments
+pub struct CompiledScript {
+    pub header: CompiledHeader,
+
+    // Below should be enough information to materialize a
+    // main<Cup<T>, W>(arg1, ..., argn) {}
+    // with args of any type (including generic types) that the VM (verification
+    // and runtime) can use.
+    // There is a verification step here that is *purely* runtime.
+    // The client is passing the type arguments and you can imagine restrictions
+    // given our resource semantic.
+
+    // function arguments signature, main is void on return
+    pub arg_types: SignatureIndex,
+    // type parameter for the script
+    pub type_params: Vec<Kind>,
+    // code
+    pub main: CodeUnit,
+}
+
+// A CompiledModule is what it used to be but you will notice the removal of `FieldDefinition`s.
+// Fields are now serialized with the `StructDefinition` which changes verification around fields.
+// (see `StructDefinition` for details)
+pub struct CompiledModule {
+    pub header: CompiledHeader,
+
+    pub struct_defs: Vec<StructDefinition>,
+    pub function_defs: Vec<FunctionDefinition>,
+}
+
+//
+// Basics...
+//
+
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+// Any expressible type in Move
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Vector(Box<Type>),
+    Struct(StructHandleIndex),
+    // a complete or partial instantiation of a generic type
+    StructInstantiation(StructInstantiationIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+    // a type argument provided by the definition (via instantiation)
+    TypeParameter(TypeParameterIndex),
+}
+
+//
+// Handles
+//
+
+
+pub struct ModuleHandle {
+    pub address: AddressPoolIndex,
+    pub name: IdentifierIndex,
+}
+
+
+pub struct StructHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub is_nominal_resource: bool,
+    pub type_params: Vec<Kind>,
+}
+
+// FunctionHandles inlines the signature as indexes into the sig table
+pub struct FunctionHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub arg_types: SignatureIndex,
+    pub return_types: SignatureIndex,
+    pub type_params: Vec<Kind>,
```

Comment:
this is `type_formals` in the current code base.
Should we rename it or leave `formals`?

Reply:
I just want to see it be consistent between values and types.  Right now it's "arg_types" and "type_params". It feels weird to me that they are args in one spot but params in another

I just want to see the same naming used for both. So some options I see for values/types
- params / type_params
- parameters / type_parameters
- arg_types / type_arg_constraints 

Something like that

Reply:
Also, I know Sam likes formals/actuals for parameters/arguments 

I don't really care, just want some consistency :)

Reply:
Agreed that consistency is the important thing :). The only reason I have a preference for formals/actuals is that we sometimes use "parameters" and "arguments" interchangeably instead of precisely. I would vote for parameters, type_parameters, and return_ (seems odd to have return_types + type_params when they are different.)

Reply:
what about `return_signature` for `return_types` instead of `return_`?
don't know... going with `return_` for now...